Fixed loss of important patch/diff aspects and too aggressive diff coloring.

sun opened the pull request:
I like the "new" diff output a lot.  However, I think we've gone a little bit too crazy with the colors — primarily the very aggressive background colors, but also the old/new font colors.

At the same time, diff context lines were changed to appear in almost invisible light gray — even though the context is of utmost importance in patch reviews.

Lastly, Dreditor already does its best to discover code comments, and it colorized them previously — doing so gives a lot more visual structure to the presented lines of code, and on top, it _encourages_ developers to (1) write comments and (2) write _good_ comments.  Especially if you're reviewing tons of patches per day/week, this visualization is key for productivity.

Changes in this PR:
- Contextual diff lines (preceding/following lines of changed lines) are crucial, must not visually appear as unimportant, greyed-out content.
- Restore colorization of comments in added lines for readability and to encourage code comments.
- Generally decrease background color saturation to shift focus on code, instead of colors.

Screenie:

![dreditor-diff-coloring](https://f.cloud.github.com/assets/41992/1475861/1ccf2d60-4642-11e3-874b-c8f1b5f0ba14.png)

sun commented:
Sometimes, a screenshot like this is even more helpful and reveals a larger picture...

In fact, when looking at this, the contextual diff lines have the same visual priority as the diff file headers - which is very obviously wrong:  The former plays a key role in patch reviews, the latter only exists to provide orientation.

markhalliwell commented:
I don't discount that the inline comments should be distinct (but also for removed lines, not just added). However, I think that should be the extent of this PR for the moment and here is why:

You think the colors are crazy. I do not. Every person has their personal preferences when it comes to visual elements. We all see better in different color schemes. I personally **cannot** effectively read light/white background based diffs and need a more prominent background to discern the differences. I actually prefer very dark based schemes (ie: black container background), which has been suggested and I turned down because I know most prefer lighter based schemes.

> Contextual diff lines (preceding/following lines of changed lines) are crucial, must not visually appear as unimportant, greyed-out content.

That is how you think. I cannot speak for everyone, but I can tell you that for me my brain works well this way. For me it's import to focus on the code that is actually being changed. I can see the grey-out content just fine, if and when I need to.

Many of these changes were introduced by neochief to mimc the "GH style" diffs. I will agree that what we have isn't the best for everyone and have tried to make some improvements here and there along the way (based on feedback/observation). Honestly though, the more and more I think about this, I realize that this issue isn't one that can be solved by attempting to change Dreditor styles. We cannot please everyone and should really move away from providing CSS in Dreditor's JS (which btw, also makes these changes absolutely unbearable IMHO).

Ultimately, I think we should instead be focusing on creating an easier way for people to save these "userstyles" (essentially) as "themes". Maybe even on dreditor.org to share/distribute? I'm not sure. Any individual could  switch these out at their convenience and load an actual stylesheet (from dreditor.org or where ever). We would simply "default" to the "theme" that we pull out of Dreditor's JS and users could customize them to their liking.

I really would prefer to get away from assuming things in Dreditor. Especially when it comes to styling and for people who say, cannot see a particular color or have colors that look similar to them? This should be where we focus our attention, not going backwards to previous styles or taking the personal preferences of either you or me.

sun commented:
Sorry, but contextual diff lines matter for technical reasons, not personal opinions.

Consider this:

``` diff
   $context = entity_get_context('node', $account, 'full');
   $name = entity_get_list_name('node', $entities);
   $list = array();
   foreach ($entities as $id => $entity) {
-    $list[] = $entity->getSummary();
+    if ($entity->isListable() && $name = $entity->getListName()) {
+      $list[$name] = $entity->getSummary();
+    }
   }
   return theme('entity_list', $name, $list);
 }
```

All wrong.

This happens, and based on my patch review experience, it happens frequently.

Regarding custom styles/themes/color-schemes, all of that is certainly a great idea, but we still have to ship with sane default styles.

The former Dreditor styles certainly were not fancy.  Instead, they were minimalistic and tailored towards simplicity and productivity.  They primarily focused on technical matters and not on visual aesthetics.

It is more than obvious that the current styles were designed to mimic GitHub — though not even GitHub uses that much of an aggressive coloring as the current Dreditor styles (code appears in black, not red/green on top of red/green background).  Furthermore, GitHub does not expose code comments.  Lastly, while I agree that GitHub serves as a good example for doing diffs, it's definitely not the best and most sane way to do it.

My primary concern is that the current color-scheme dominates too much.  It distracts from the actual code (it distracts from the actual purpose), and some crucial aspects for patch reviews were unfortunately lost.  As a core developer who reviewed up to ~100 patches per day, these functional/technical aspects did play a key role.  We want to ensure the app solves the problem it was designed for.

Overall, I guess I'm missing concrete feedback on which parts you (or actually, all other users) agree, and on which parts you disagree, and, whether we can find sensible compromises for those questions.

E.g., if it's just that added lines should appear in green (as opposed to blue), then no problem at all, we just need to rethink coloring of comments (black? blue?).

So please don't get me wrong, I'm all open for compromises; the primary goal is to restore functional/technical aspects for patch reviews.

markhalliwell commented:
> Sorry, but contextual diff lines matter for technical reasons, not personal opinions.

Core is different from contrib. When you're working on smaller patches in code one knows, these contextual lines _can_ sometimes become annoying and not relevant. That being said, I'm not saying that they are irrelevant. I'm ok with changing them to `#777`. Anything darker though honestly pulls the focus from the added/removed line, for me anyway. Perhaps my eyes need more contrast/depth (between context/add/remove lines, not talking about foreground/background of text here) than others. I darkened the image below to `#777` btw to see if this is an acceptable compromise.

> My primary concern is that the current color-scheme dominates too much. It distracts from the actual code (it distracts from the actual purpose), and some crucial aspects for patch reviews were unfortunately lost.

I'm sure it distracts you, that is my point. It does not distract me one bit. In fact, it is quite the opposite. It actually helps me scan the code quickly before looking at it in more detail.

> Overall, I guess I'm missing concrete feedback on which parts you (or actually, all other users) agree, and on which parts you disagree, and, whether we can find sensible compromises for those questions.

I am very visual and have a somewhat photographic memory. Having darker "chunks" of red or green allows me to scroll rather quickly through-out a patch because my brain can create mental landmarks based on the size and colors of these chunks. This is the very first thing I do when opening any patch. I do not have to read the text to actually know where I am. If we simply focused the color on the text foreground, there isn't enough contrast (between background/foreground this time) for my brain/eyes to accomplish this task. This is why I, personally, need darker and more distinct "chunks" to quickly scan.

To give you an example of how my brain parses the following:
![](https://www.monosnap.com/image/afIEC5yOm4GnMhEVKMQOu4x6O.png)

I see that there are two red chunks at the top of approximately 4-6 lines each. I then see the additions below are individual lines with a chunk of green before it. I then look at the code in more detail reading the context to gain perspective and then proceed to read the additions and deletions from top to bottom. I then associate the actual value of this section to it's "section signature" (if you will).

I then would move on to the rest of the patch, but something causes me to scroll back up the this particular section because it's related. By having first scanned the patch I can quickly identify _exactly_ where it is that I need to scroll to: 2 small red, 1 small green, 1 single green and a replacement (single red/single green).

The whole point I was trying to make is that we honestly shouldn't go one way or the other based on personal preferences (which is exactly what this is). No human processes information the same way. We have to move forward and allow easier customization based on individual preferences.

> E.g., if it's just that added lines should appear in green (as opposed to blue), then no problem at all, we just need to rethink coloring of comments (black? blue?).

I think yes, we should have some fundamental agreement on colors for what they represent in the patch, which is what I have tried to do:

Additions: green
Removals: red
Context: [darker] grey
Headers: light blue
Selection: light yellow

Re: comments in lines
PHP colors comments as `#FF8000`. I like this, but it doesn't work for removed lines. I think we'll need to change based on whether they're in context, additions or removals. Black isn't really viable because it would be darker than everything else and completely stick out like a sore thumb. Maybe for add/remove lines we could do their muted complementary color or simply dark grey? Also we need to actually fix detecting and wrapping comments properly, they don't always get parsed.

YesCT commented:
I agree with "sane default styles."

having context lines darker than the git diff header stuff is good to help differentiate them.

It would be cool to have comments colored different in both the removed and added lines, but looks like we .. dont know what those colors might be that would work for that? How about keeping the green text for added lines, but blue (or bluish/greenish) for added comment lines. If we want to do the same for removed, maybe red for removed code and redish/purplish for removed comments?

I'm not sure what "PHP colors comments" means... is it referring to standard php syntax highlighting, or a particular editor used for php?

for the green on green and red on red... can we use accessibility guidance here to decide how light the background should be directly under the code so there is enough contrast? (and leave it darker on the sides)

dreditor "help" to save user settings for styling would be a separate issue, right? I'd like user styles to be hunkable. so I could pick one person's review styling and another person's issue queue issue edit page styling.

sun committed:
Fixed loss of important patch/diff aspects and too aggressive diff coloring.

- Contextual diff lines (preceding/following lines of changed lines) are crucial, must not visually appear as unimportant, greyed-out content.
- Restore colorization of comments in added lines for readability and to encourage code comments.
- Generally decrease background color saturation to shift focus on code, instead of colors.

sun committed:
Fixed loss of important patch/diff aspects and too aggressive diff coloring.

- Contextual diff lines (preceding/following lines of changed lines) are crucial, must not visually appear as unimportant, greyed-out content.
- Restore colorization of comments in added lines for readability and to encourage code comments.
- Generally decrease background color saturation to shift focus on code, instead of colors.
